Datadog SDK OTel Metrics and Logs API Support: Phase 1 (#32610)
DataDog/documentation: content/en/opentelemetry/instrument/api_support/dotnet/logs.md

---
title: .NET OpenTelemetry Logs Support
code_lang: logs
type: multi-code-lang
code_lang_weight: 2
further_reading:
    - link: opentelemetry/correlate/logs_and_traces
      tag: Documentation
      text: Correlate OpenTelemetry Traces and Logs
---

## Overview

You can export OTLP logs from your .NET application using the Datadog SDK (`dd-trace-dotnet`).

This feature works by intercepting logs from the built-in `Microsoft.Extensions.Logging.ILogger` library and exporting them as structured OTLP data.

> <div class="alert alert-info">This implementation does not use the OpenTelemetry Logs API. It is not compatible with the <code>OpenTelemetry.SDK.Logs</code> or <code>OpenTelemetry.Exporter.Otlp.Logs</code> packages. You must remove them to use this feature</div>

Reviewer: ```suggestion
<div class="alert alert-info">This implementation does not use the OpenTelemetry Logs API. It is not compatible with the <code>OpenTelemetry.SDK.Logs</code> or <code>OpenTelemetry.Exporter.Otlp.Logs</code> packages. You must remove them to use this feature.</div>
```

## Prerequisites

- **Datadog SDK**: `dd-trace-dotnet` version [3.30.0][4] or later.
- **An OTLP-compatible destination**: You must have a destination ready to receive OTLP data, such as the Datadog Agent or OpenTelemetry Collector.

## Setup

1. Install the Datadog SDK. Follow the installation steps for your runtime:
   - [.NET Framework][2]
   - [.NET Core][3]
2. Enable OTel logs export by setting the following environment variable:
    ```sh
    export DD_LOGS_OTEL_ENABLED=true
    ```

## Examples

### Standard logging

```csharp
[TODO: Code example]
```

> ### Trace and log correlation

Reviewer 2: I think we can also link to the traces page for the "OTel Tracing API support" text.

```suggestion
### Trace and log correlation

This example shows how logs emitted within an active Datadog span are automatically correlated. If you are using the OTel Tracing API or built-in .NET Activity API to create spans, ensure OTel Tracing API support is enabled.
```

```csharp
[TODO: Code example]
```

### Trace and log correlation

```csharp
[TODO: Code example]
```

## Supported configuration

To enable this feature, you must set `DD_LOGS_OTEL_ENABLED=true`.

All OTLP exporter settings (such as endpoints, protocols, and timeouts), resource attributes, and batch processor settings are configured using a shared set of OpenTelemetry environment variables.

For a complete list of all shared OTLP environment variables, see [OpenTelemetry Environment Variables Interoperability][1].

## Migrate from other setups

### Existing OTel setup (manual configuration)

> This Datadog feature is not compatible with applications already using the OpenTelemetry Logs API or OpenTelemetry SDK for logs.

Reviewer: ```suggestion
This Datadog feature is not compatible with applications already using the OpenTelemetry Logs API or OpenTelemetry SDK for logs.
```
This seems like a fairly important callout, like if you are using X or Y, do not try these steps, may be worth converting to a Note callout ?

---
title: .NET OpenTelemetry Logs Support
code_lang: logs
type: multi-code-lang
code_lang_weight: 2
further_reading:
    - link: opentelemetry/correlate/logs_and_traces
      tag: Documentation
      text: Correlate OpenTelemetry Traces and Logs
---

## Overview

You can export OTLP logs from your .NET application using the Datadog SDK (`dd-trace-dotnet`).

This feature works by intercepting logs from the built-in `Microsoft.Extensions.Logging.ILogger` library and exporting them as structured OTLP data.

> <div class="alert alert-info">This implementation does not use the OpenTelemetry Logs API. It is not compatible with the <code>OpenTelemetry.SDK.Logs</code> or <code>OpenTelemetry.Exporter.Otlp.Logs</code> packages. You must remove them to use this feature.</div>

Reviewer 2: I think the "Native" callout is also sufficient here
```suggestion
<div class="alert alert-info">You should not install the official OpenTelemetry SDK or any OTLP Exporter packages. The Datadog SDK provides this functionality. Installing both can lead to runtime conflicts and duplicate data.</div>
```

This Datadog feature is not compatible with applications already using the OpenTelemetry Logs API or OpenTelemetry SDK for logs.

To migrate, you must remove the OpenTelemetry logging packages (such as `OpenTelemetry.SDK.Logs` or `OpenTelemetry.Exporter.Otlp.Logs`) and any code that manually configures or uses the OTel Logs API.

This feature works by intercepting logs directly from the `Microsoft.Extensions.Logging.ILogger` interface and does not interact with the OpenTelemetry Logs API.

### Existing Datadog log injection

If you are using Datadog's traditional log injection (where `DD_LOGS_INJECTION=true` adds trace context to text logs) and an Agent to tail log files:

1. Set the `DD_LOGS_OTEL_ENABLED=true` environment variable.
2. The Datadog SDK automatically disables the old log injection style (`DD_LOGS_INJECTION`) to prevent duplicate trace metadata in your logs. Trace correlation is handled by the structured OTLP payload.
3. Ensure your Datadog Agent is configured to receive OTLP logs (version 7.48.0 or greater is required)
4. Disable any file-based log collection for this service to avoid duplicate logs.

## Troubleshooting

> {{% otel-api-troubleshooting signal="logs" %}}

Reviewer 2: I wonder if we could do a list "start" and "end" snippet so we can add additional items into the same `<ul>` HTML section. WDYT?

> - Verify the Datadog Profiler is attached. This feature relies on Datadog's automatic instrumentation. Ensure the .NET Profiler is correctly enabled for your application.

Reviewer 2: Same comment as https://github.com/DataDog/documentation/pull/32610/files#r2516341501 . "Datadog Profiler" can be confusing. Can we reword so it's not confused with the Datadog Continuous Profiler product?

1.  Install the Datadog SDK:
    ```sh
    dotnet add package Datadog.Trace
    ```
2.  Enable OTel logs export by setting the following environment variable:
    ```sh
    export DD_LOGS_OTEL_ENABLED=true
    ```
3.  Ensure Datadog instrumentation is active for your application.
    ```sh
    Command for enabling .NET profiler
    ```

## Examples

### Standard logging

### Trace and log correlation

## Supported configuration

To enable OTLP Logs support, you must set `DD_LOGS_OTEL_ENABLED=true`.

While Datadog supports many OpenTelemetry variables, these are the most common ones needed for a basic setup.

`DD_LOGS_OTEL_ENABLED`
: **Description**: Enables the Datadog SDK to intercept standard logs and export them in OTLP format.
: **Default**: `false`

`OTEL_EXPORTER_OTLP_LOGS_PROTOCOL`
: **Description**: Specifies the OTLP transport protocol. Takes precedence over `OTEL_EXPORTER_OTLP_PROTOCOL`.
: **Accepted values**: `grpc`, `http/protobuf`, `http/json`
: **Default**: (SDK-dependent)

`OTEL_EXPORTER_OTLP_LOGS_ENDPOINT`
: **Description**: Specifies the URL for sending OTLP logs. Takes precedence over `OTEL_EXPORTER_OTLP_ENDPOINT`.
: **Default (gRPC)**: `http://localhost:4317`

> : **Default (HTTP)**: `http://localhost:4318/v1/logs`

Reviewer 3: These configurations will be used in all otel log implementations. Should we move the definitions to the content/en/opentelemetry/config/environment_variable_support.md page?
